pkg/liquidity-mining/contracts/fee-distribution/FeeDistributor.sol

```diff
@@ -0,0 +1,369 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/IAuthentication.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+
+import "../interfaces/IVotingEscrow.sol";
+
+contract FeeDistributor is ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    uint256 private constant _TOKEN_CHECKPOINT_DEADLINE = 1 days;
+
+    IVotingEscrow private immutable _votingEscrow;
+
+    uint256 private immutable _startTime;
+
+    // Global State
+    uint256 private _timeCursor;
+    mapping(uint256 => uint256) private _veSupplyCache;
+
+    // Token State
+    mapping(IERC20 => uint256) private _tokenStartTime;
+    mapping(IERC20 => uint256) private _tokenTimeCursor;
+    mapping(IERC20 => uint256) private _tokenLastBalance;
+    mapping(IERC20 => mapping(uint256 => uint256)) private _tokensPerWeek;
+
+    // User State
+    mapping(address => uint256) private _userTimeCursor;
+    mapping(address => uint256) private _userLastEpochCheckpointed;
+    mapping(address => mapping(uint256 => uint256)) private _userBalanceAtTimestamp;
+    mapping(address => mapping(IERC20 => uint256)) private _userTokenTimeCursor;
```

Comment:
Several of these mappings should be replaced with mappings to packed structs, however this should be done after initial reviews for readability sake.

Reply:
I've got these changes in a separate branch so will make a PR after this one to apply these changes.